### ISSUE TITLE
gather_facts is not run when play is tagged and different tag specified

### ISSUE THREAD

**Issue by pgassmann** [ansible/ansible#60833]
##### SUMMARY

Facts are not gathered when using `--tags my_tag` if play is tagged with a different tag.

##### ISSUE TYPE
- Bug Report
- Regression

##### ANSIBLE VERSION

Affects `ansible>=2.8.0` incl. current latest version `ansible==2.8.4`

works as expected with `ansible==2.7.1`

##### CONFIGURATION

```paste below
DEFAULT_GATHERING(ansible.cfg) = smart
```

##### STEPS TO REPRODUCE

Playbook:
```yaml
- name: Gather facts with tags
  hosts: localhost
  gather_facts: yes
  tasks:
    - debug:
        var: ansible_domain
      tags:
        - my_tag
  tags:
    - not_my_tag
```
run with `ansible-playbook test.yml --tags my_tag`

##### EXPECTED RESULTS

```bash
$ ansible-playbook test.yml --tags my_tag
PLAY [Gather facts with tags] ************************************

TASK [Gathering Facts] *************************************
ok: [localhost]

TASK [debug] *****************************************
ok: [localhost] => 
  ansible_domain: 'example.com'

PLAY RECAP *****************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

##### ACTUAL RESULTS

```bash
$ ansible-playbook test.yml --tags my_tag

PLAY [Gather facts with tags] *****************************************

TASK [debug] *****************************************
ok: [localhost] => 
  ansible_domain: VARIABLE IS NOT DEFINED!

PLAY RECAP *****************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

##### WITHOUT TAGS ON PLAY
Works as expected. `TASK [Gathering Facts]` is run.

```yaml
- name: Gather facts with tags
  hosts: localhost
  gather_facts: yes
  tasks:
    - debug:
        var: ansible_domain
      tags:
        - my_tag
  ## no tags on play
  # tags: 
  #   - not_my_tag
```
Result:

```bash
$ ansible-playbook test.yml --tags my_tag
PLAY [Gather facts with tags] ************************************

TASK [Gathering Facts] *************************************
ok: [localhost]

TASK [debug] *****************************************
ok: [localhost] => 
  ansible_domain: 'example.com'

PLAY RECAP *****************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

**Comment by ansibot** [comment]
Files identified in the description:
None

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by resmo** [comment]
bisected to 4ec2423a049551733f0f6a5d7be5202eb4a02d27

**Comment by resmo** [comment]
I see the "benefit" in the hidden always tag on facts gathering as well. 
The question is, how users could enable the always tag back on facts gathering or configure this behavior as "default"? The fix in #44251 hard breaks it. 

I would suggest to add a new feature: A possible solution would be a new keyword for facts gathering e.g. `gather_facts: always-yes`, a `always-`preifx to make sure we want to always run in on tags:

```yaml
- name: Gather facts with tags
  hosts: localhost
  gather_facts: always-yes # always-smart, etc...
  tasks:
    - debug:
        var: ansible_domain
      tags:
        - my_tag
  # no tags on play
  tags: 
    - not_my_tag
```

another option I see would  be to allow tags on facts gathering  "task"

```yaml
- name: Gather facts with tags
  hosts: localhost
  gather_facts: yes
    tags:
      - my_tag
  tasks:
    - debug:
        var: ansible_domain
      tags:
        - my_tag
  # no tags on play
  tags: 
    - not_my_tag
```

My vote would be the 2nd, as it seems to be the obvious solution to users.

**Comment by willthames** [comment]
@resmo - we'd have to document such an option anyway so my preferences would be:

* always run `gather_facts` if it's set, no matter what tags are given
* document use of
   ```
  tasks:
    - setup:
      tags:
        - always
  ```

**Comment by willthames** [comment]
One other point is that 
```
  gather_facts: yes
    tags:
      - my_tag
```
is not valid yaml - gather_facts would need to be of type `bool/str/dict` (I think there are precedences for that) and you could just have:
```
  gather_facts:
    tags:
      - my_tag
```

**Comment by sivel** [comment]
Duplicate of https://github.com/ansible/ansible/issues/58867

**Comment by bcoca** [comment]
@resmo, @willthames @sivel, from previous ticket i had already thought of adding settting `'gathering_always_runs=no(default)|yes` (bikeshed away) to allow for the previous behaviour for those that want it.

**Comment by pgassmann** [comment]
@bcoca What was the scenario and "bug" that the "bugfix/regression" tries to fix? I expect that if one can run a specific task in a play by specifying a tag, it should also have the facts available unless `gather_facts` is set to false?

**Comment by bcoca** [comment]
the scenario was that 'gather_facts' was ignoring tags, so when someone did `--skip-tags this` on a play tagged with 'this' it was executing fact gathering when user did not want it to, at the time this clearly looked like a bug.  Only after release did people seem to expect and even want this behaviour, so now a toggle seems to be the only way to satisfy both groups.
